internal/testtypes/matchers.go

```diff
@@ -0,0 +1,275 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testtypes
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/onsi/gomega/format"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// This code is a modified version of the mergePatch code from the controllers/topology/internal/mergepatch pkg.
+// TODO: Create a shared patchdiff library that can be shared for both patching and matching.
+
+// These package variables hold pre-created commonly used options that can be used to reduce the manual work involved in
+// identifying the paths that need to be compared for testing equality between objects.
+var (
+
+	// IgnoreAutogeneratedMetadata contains the paths for all the metadata fields that are commonly set by the
+	// client and APIServer. This is used as a MatchOption for situations when only user-provided metadata is relevant.
+	IgnoreAutogeneratedMetadata = IgnorePaths{
+		{"metadata", "uid"},
+		{"metadata", "generation"},
+		{"metadata", "creationTimestamp"},
+		{"metadata", "resourceVersion"},
+		{"metadata", "managedFields"},
+		{"metadata", "deletionGracePeriodSeconds"},
+		{"metadata", "deletionTimestamp"},
+		{"metadata", "ownerReferences"},
+		{"metadata", "selfLink"},
+		{"metadata", "generateName"},
+	}
+)
+
+// Matcher is a Gomega matcher used to establish equality between two Kubernetes runtime.Objects.
+type Matcher struct {
+	// original holds the object that will be used to Match.
+	original runtime.Object
+
+	// diff contains the delta between the two compared objects.
+	diff []byte
+
+	// options holds the options that identify what should and should not be matched.
+	options *MatchOptions
+}
+
+// EqualObject applies the MatchOptions and returns a Matcher for the passed Kubernetes runtime.Object. This function
```

Comment:
I'm not sure I would include `applies the MatchOptions` it almost sounds like we apply them to the object. I probably would just omit it.